### PR TITLE
[BE-4558] Adds support for USING in delete

### PR DESCRIPTION
**Pull request by keatoooon**
@allai5 @jsternberg , whichever of you have time, please review.

### CODE REVIEW

File: core/src/main/codegen/templates/Parser.jj [hunk 1]
```diff
@@ -1728,10 +1730,65 @@ SqlNode SqlDelete() :
         }
     ]
     [ [ <AS> ] alias = SimpleIdentifier() ]
+    [ <USING> using = SqlDeleteUsingList() ]
```
Comment (keatoooon) [inline core/src/main/codegen/templates/Parser.jj:1734]: Note that all of the parsing changes will need to be ported to bodo's parser

File: core/src/main/java/org/apache/calcite/sql/SqlCall.java [hunk 2]
```diff
@@ -140,7 +140,7 @@ public int operandCount() {
     final SqlDialect dialect = writer.getDialect();
     if (leftPrec > operator.getLeftPrec()
         || (operator.getRightPrec() <= rightPrec && (rightPrec != 0))
-        || writer.isAlwaysUseParentheses() && isA(SqlKind.EXPRESSION)) {
+        || (writer.isAlwaysUseParentheses() && isA(SqlKind.EXPRESSION))) {
```
Comment (keatoooon) [inline core/src/main/java/org/apache/calcite/sql/SqlCall.java:143]: There is an open TODO to bug the calcite people about this, but based on my understanding, this is the correct grouping

File: core/src/main/java/org/apache/calcite/sql/SqlDelete.java [hunk 3]
```diff
@@ -44,12 +46,14 @@ public class SqlDelete extends SqlCall {
   public SqlDelete(
       SqlParserPos pos,
       SqlNode targetTable,
+      @Nullable SqlNodeList usingList,
```
Comment (jsternberg) [inline core/src/main/java/org/apache/calcite/sql/SqlDelete.java]: Can you add a new constructor to this or utilize a setter to set this value? Just so we don't have to change the invocation of this function in other places.

File: core/src/main/java/org/apache/calcite/sql/SqlWithItem.java [hunk 4]
```diff
@@ -92,10 +92,12 @@ private static class SqlWithItemOperator extends SqlSpecialOperator {
         int leftPrec,
         int rightPrec) {
       final SqlWithItem withItem = (SqlWithItem) call;
+
```
Comment (jsternberg) [inline core/src/main/java/org/apache/calcite/sql/SqlWithItem.java]: Random formatting changes. Please remove since this isn't a file we're directly touching.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 5]
```diff
@@ -1480,8 +1482,14 @@ private void handleOffsetFetch(@Nullable SqlNode offset, @Nullable SqlNode fetch
 
     case DELETE: {
       SqlDelete call = (SqlDelete) node;
-      SqlSelect select = createSourceSelectForDelete(call);
-      call.setSourceSelect(select);
+      if (call.getUsing() != null) {
+        //If we have a USING clause, we rewrite the delete as a merge operation
```
Comment (jsternberg) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Comment should have a space after them. So `// If`.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 6]
```diff
@@ -1704,6 +1735,88 @@ private static void rewriteMerge(SqlMerge call) {
...
+
+    SqlDelete matchedDeleteExpression = new
+        SqlDelete(originalDeleteCall.getParserPosition(),
+        targetTable, null, null, null, alias);
+
+    //There's a wierd issue here. Essentially, in validation, it will infer
+    // the row type of EMP as the row type from the merge in one location, but
+    // as the emp from the original table in another location, which causes a number conflicts.
+    //
+    // I spent about a day trying to resolve this, but all the fixes I tried ended up breaking
```
Comment (jsternberg) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:1783]: Put `TODO(keaton)` here on this note.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 7]
```diff
@@ -1770,7 +1883,7 @@ private SqlNode rewriteUpdateToMerge(
     source = SqlValidatorUtil.addAlias(source, UPDATE_SRC_ALIAS);
     SqlMerge mergeCall =
         new SqlMerge(updateCall.getParserPosition(), target, condition, source,
-            SqlNodeList.of(updateCall), SqlNodeList.EMPTY, null,
+            SqlNodeList.of(updateCall), SqlNodeList.EMPTY.clone(SqlParserPos.ZERO), null,
```
Comment (jsternberg) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Can you explain why this change?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 8]
```diff
@@ -1875,6 +1988,7 @@ protected SqlSelect createSourceSelectForUpdate(SqlUpdate call) {
    * @return select statement
    */
   protected SqlSelect createSourceSelectForDelete(SqlDelete call) {
+
```
Comment (jsternberg) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Unnecessary formatting change.

**Review by jsternberg — CHANGES_REQUESTED** [review]
Mostly fine. A few formatting concerns and some questions.

File: core/src/main/codegen/templates/Parser.jj [hunk 9]
```diff
@@ -1728,10 +1730,65 @@ SqlNode SqlDelete() :
...
+SqlDeleteUsingItem SqlDeleteUsingElement() :
+{
+    SqlNode deleteUsingElement;
+    SqlIdentifier alias = null;
+    SqlParserPos s;
+}
+{
+    // Based on my reading of with parsing, this should allow
+    // for either a table identifier, or
+    // any subquery valid in a "with" cluase (which is the expected
```
Comment (allai5) [inline core/src/main/codegen/templates/Parser.jj]:
```suggestion
    // any subquery valid in a WITH clause (which is the expected
```

File: core/src/main/java/org/apache/calcite/sql/SqlCall.java (hunk 2 above)
Comment (allai5) [reply core/src/main/java/org/apache/calcite/sql/SqlCall.java:143]: Is there a test that is affected by this change? And if so can you explain this grouping in a bit more detail for my learning

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 10]
```diff
@@ -1480,8 +1482,14 @@ private void handleOffsetFetch(@Nullable SqlNode offset, @Nullable SqlNode fetch
 
     case DELETE: {
       SqlDelete call = (SqlDelete) node;
-      SqlSelect select = createSourceSelectForDelete(call);
-      call.setSourceSelect(select);
+      if (call.getUsing() != null) {
+        //If we have a USING clause, we rewrite the delete as a merge operation
+        node = rewriteDeleteToMerge(call);
+      } else {
+        //Otherwise, we leave it as is, and just generate the source select
```
Comment (allai5) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Same here for space after comment.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 11]
```diff
@@ -1583,14 +1591,37 @@ private static void rewriteMerge(SqlMerge call) {
...
           SqlValidatorUtil.addAlias(
               targetTable,
               call.getAlias().getSimple());
+    } else {
+      // Due to the manner in which calcite handles subqueries, we need to explicitly add an
+      // alias here in order to handle fully qualification.
+      //
+      // For example this will succseed:
+      //
+      // SELECT *, True from dept inner join (SELECT *, True from emp) as emp on emp.ename = 1
```
Comment (allai5) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:1608]: Add carats (^^^^) to the difference between the two examples.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 12]
```diff
@@ -1583,14 +1591,37 @@ private static void rewriteMerge(SqlMerge call) {
...
+        origTargetTable, null, null, null, null, null,
         null, null, null, null);
     if (call.getAlias() != null) {
       targetTable =
           SqlValidatorUtil.addAlias(
               targetTable,
               call.getAlias().getSimple());
+    } else {
+      // Due to the manner in which calcite handles subqueries, we need to explicitly add an
+      // alias here in order to handle fully qualification.
```
Comment (allai5) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]:
```suggestion
      // alias here in order to handle fully qualified table names.
```

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 13]
```diff
@@ -1704,6 +1735,88 @@ private static void rewriteMerge(SqlMerge call) {
 
   }
 
+
+
+  /**
+   * Used in UnconditonalRewrites. In the case that we have a delete with a using clause.
```
Comment (allai5) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]:
```suggestion
   * Used in UnconditonalRewrites. In the case that we have a DELETE with a USING clause.
```

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 14]
```diff
@@ -1704,6 +1735,88 @@ private static void rewriteMerge(SqlMerge call) {
...
+   *
+   * DELETE FROM target using T1, T2, ... where (cond) is equivalent to
+   *
+   * MERGE INTO target using (T1 full outer join T2...) on (cond) WHEN MATCHED THEN DELETE
+   *
+   * We choose to do this rewrite (similar to rewriteUpdateToMerge), in order to simplify validation
+   * and sqlToRel code generation.
+   *
+   * @param originalDeleteCall The delete call to transform. Must have at least one table/subquery
```
Comment (allai5) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]:
```suggestion
   * @param originalDeleteCall The DELETE call to transform. Must have at least one table/subquery
```

File: core/src/main/java/org/apache/calcite/sql/SqlCall.java (hunk 2 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/SqlCall.java:143]: So, essentially, && has higher precedence then ||, so the addition of parenthesis here:
`(writer.isAlwaysUseParentheses() && isA(SqlKind.EXPRESSION)`

Doesn't change anything. However, based on my understanding of the overall logic, I think the correct grouping should be around the first three expressions as so:

`(leftPrec > operator.getLeftPrec()
        || (operator.getRightPrec() <= rightPrec && (rightPrec != 0))
        || (writer.isAlwaysUseParentheses()) && isA(SqlKind.EXPRESSION)`

Because it doesn't make sense to put parenthesis around something that isn't an expression. However, I wanted to bug the Calcite people first, just in case there is something that I'm not seeing.

File: core/src/main/java/org/apache/calcite/sql/SqlDelete.java (hunk 3 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/SqlDelete.java]: Done

File: core/src/main/java/org/apache/calcite/sql/SqlWithItem.java (hunk 4 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/SqlWithItem.java]: Done

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 10 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Done

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 11 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:1608]: Done.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 13 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: done

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 14 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: done

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 6 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:1783]: done

**Review by keatoooon — COMMENTED** [review]
Made suggested changes

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 7 above)
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: I had an issue where I was appending to SqlNodeList.EMPTY somewhere.

This was leftover code from debugging, it should have been reverted.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 15]
```diff
@@ -1704,6 +1735,89 @@ private static void rewriteMerge(SqlMerge call) {
...
+    }
+    SqlIdentifier alias = originalDeleteCall.getAlias();
+
+    SqlNodeList matchedCallList = SqlNodeList.EMPTY.clone(originalDeleteCall.getParserPosition());
+
+    SqlDelete matchedDeleteExpression = new
+        SqlDelete(originalDeleteCall.getParserPosition(),
+        targetTable, null, null, alias);
+
+    //There's a wierd issue here. Essentially, in validation, it will infer
```
Comment (jsternberg) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Misspelled "weird". Also formatting with the `//`.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 16]
```diff
@@ -1704,6 +1735,89 @@ private static void rewriteMerge(SqlMerge call) {
...
+    // the row type of EMP as the row type from the merge in one location, but
+    // as the emp from the original table in another location, which causes a number conflicts.
+    //
+    // I spent about a day trying to resolve this, but all the fixes I tried ended up breaking
+    // existing merge into paths, and we don't even use the created Delete List anywhere
+    // in the merge path, I'm just going to set this to something arbitrary, and leave this as
+    // technical debt to figure out later.
+    // TODO(keaton)
```
Comment (jsternberg) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: The todo should go before the message.

**Review by jsternberg — APPROVED** [review]
Approving. Just do a search for `//` without a space after it and fix those up. The content looks good so I'm not going to block merging on cosmetic changes.

**Comment by keatoooon** [comment]
Searched for un-indented comments in my changes via the following regex: "\s//[^\s]". Might have missed a few, since there was a lot of output.
